File: packages/typescript/src/resources/shodan.ts
import type { Transport } from "../client.js";
import type {
  ShodanCustomScanRequest,
  ShodanScan,
  ShodanScanList,
  ShodanHost,
  ResourceId,
} from "../types.js";

export class ShodanResource {
  readonly #transport: Transport;
  constructor(transport: Transport) {
    this.#transport = transport;
  }

  customScans(): Promise<ShodanScanList> {
    return this.#transport.request<ShodanScanList>("GET", "/api/v1/shodan/custom-scans");
  }

  customScan(body: ShodanCustomScanRequest): Promise<ShodanScan> {
    return this.#transport.request<ShodanScan>("POST", "/api/v1/shodan/custom-scan", { body });
  }

  getScan(id: ResourceId): Promise<ShodanScan> {
    return this.#transport.request<ShodanScan>("GET", `/api/v1/shodan/custom-scans/${encodeURIComponent(String(id))}`);
  }

  deleteScan(id: ResourceId): Promise<unknown> {
    return this.#transport.request<unknown>("DELETE", `/api/v1/shodan/custom-scans/${encodeURIComponent(String(id))}`);
  }

  host(ip: string): Promise<ShodanHost> {
    return this.#transport.request<ShodanHost>("GET", `/api/v1/shodan/host/${encodeURIComponent(ip)}`);
  }
}

File: rust/tests/shodan.rs
use leaksnow::{Client, ShodanCustomScanRequest};
use wiremock::matchers::{method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};

#[tokio::test]
async fn custom_scans_get() {
    let server = MockServer::start().await;
    Mock::given(method("GET")).and(path("/api/v1/shodan/custom-scans"))
        .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!([])))
        .expect(1).mount(&server).await;
    let c = Client::builder("ms_key").base_url(server.uri()).build();
    c.shodan().custom_scans().await.unwrap();
}

#[tokio::test]
async fn custom_scan_post() {
    let server = MockServer::start().await;
    Mock::given(method("POST")).and(path("/api/v1/shodan/custom-scan"))
        .and(wiremock::matchers::body_json(serde_json::json!({"target": "scanme.nmap.org"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({"id": 1})))
        .expect(1).mount(&server).await;
    let c = Client::builder("ms_key").base_url(server.uri()).build();
    c.shodan().custom_scan(ShodanCustomScanRequest { target: "scanme.nmap.org".into() }).await.unwrap();
}

#[tokio::test]
async fn get_delete_scan_and_host() {
    let server = MockServer::start().await;
    Mock::given(method("GET")).and(path("/api/v1/shodan/custom-scans/42"))
        .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({}))).mount(&server).await;
    Mock::given(method("DELETE")).and(path("/api/v1/shodan/custom-scans/42"))
        .respond_with(ResponseTemplate::new(200)).mount(&server).await;
    Mock::given(method("GET")).and(path("/api/v1/shodan/host/1.2.3.4"))
        .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!({}))).mount(&server).await;

    let c = Client::builder("ms_key").base_url(server.uri()).build();
    c.shodan().get_scan("42").await.unwrap();
    c.shodan().delete_scan("42").await.unwrap();
    c.shodan().host("1.2.3.4").await.unwrap();
}
